src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/haflows/HaFlowHistorySpec.groovy

```diff
@@ -0,0 +1,258 @@
+package org.openkilda.functionaltests.spec.flows.haflows
+
+import static org.openkilda.functionaltests.extension.tags.Tag.LOW_PRIORITY
+import static org.openkilda.testing.Constants.WAIT_OFFSET
+import static org.openkilda.testing.service.floodlight.model.FloodlightConnectMode.RW
+
+import org.openkilda.functionaltests.HealthCheckSpecification
+import org.openkilda.functionaltests.error.HistoryMaxCountExpectedError
+import org.openkilda.functionaltests.extension.failfast.Tidy
+import org.openkilda.functionaltests.extension.tags.Tags
+import org.openkilda.functionaltests.helpers.Wrappers
+import org.openkilda.functionaltests.helpers.model.HaFlowExtended
+import org.openkilda.messaging.info.event.IslChangeType
+import org.openkilda.messaging.payload.flow.FlowState
+import org.openkilda.messaging.payload.history.HaSubFlowPayload
+import org.openkilda.model.FlowEncapsulationType
+import org.openkilda.model.SwitchId
+import org.openkilda.model.history.DumpType
+import org.openkilda.northbound.dto.v2.flows.BaseFlowEndpointV2
+import org.openkilda.northbound.dto.v2.haflows.HaFlowPatchPayload
+import org.openkilda.testing.service.northbound.model.HaFlowActionType
+
+import org.springframework.web.client.HttpClientErrorException
+import spock.lang.Narrative
+
+@Narrative("""Verify that history records are created for the basic actions applied to Ha-Flow.""")
+class HaFlowHistorySpec extends HealthCheckSpecification {
+
+    @Tidy
+    def "History records with links details are created during link #type operations and can be retrieved with or without timeline"() {
+        given: "HA-Flow has been created"
+        def swT = topologyHelper.switchTriplets[0]
+        Long timeBeforeOperation = System.currentTimeSeconds()
+        HaFlowExtended haFlow = HaFlowExtended.build(swT, northboundV2, topology).create()
+
+        when: "#type action has been executed"
+        action(haFlow)
+
+        then: "Correct event appears in HA-Flow history and can be retrieved without specifying timeline"
+        def historyRecordWithoutTimeLine = haFlow.getHistory().getEntriesByType(type)
+//        https://github.com/telstra/open-kilda/issues/5320
+        if (type != HaFlowActionType.UPDATE) {
+            assert historyRecordWithoutTimeLine.size() == 1
+        }
+
+        verifyAll {
+            historyRecordWithoutTimeLine[0].haFlowId == haFlow.haFlowId
+            historyRecordWithoutTimeLine[0].taskId
+            historyRecordWithoutTimeLine[0].timestampIso
+
+//            https://github.com/telstra/open-kilda/issues/5366
+//            expectedHistoryRecord[0].payloads.action.find {it == type.getPayloadLastAction()}
+            historyRecordWithoutTimeLine[0].payloads.every { it.timestampIso }
+        }
+
+        and: "Flow history contains all flow properties in the dump section"
+        verifyAll {
+            with(historyRecordWithoutTimeLine.first().dumps[0]) { dump ->
```

Comment:
May be it's better to add method HaFlowHistoryEntry.matches(HaFlowExtended haFlow)?

Reply:
Decided to leave this section of verification inside the spec as it is only one place where it is used.